include/deal.II/fe/fe_interface_values.h

```diff
@@ -2242,35 +2412,83 @@ FEInterfaceValues<dim, spacedim>::reinit(
   const unsigned int              face_no_neighbor,
   const unsigned int              sub_face_no_neighbor)
 {
-  if (sub_face_no == numbers::invalid_unsigned_int)
-    {
-      internal_fe_face_values.reinit(cell, face_no);
-      fe_face_values = &internal_fe_face_values;
-    }
-  else
+  if (internal_fe_face_values)
     {
-      internal_fe_subface_values.reinit(cell, face_no, sub_face_no);
-      fe_face_values = &internal_fe_subface_values;
+      if (sub_face_no == numbers::invalid_unsigned_int)
+        {
+          internal_fe_face_values->reinit(cell, face_no);
+          fe_face_values = internal_fe_face_values.get();
+        }
+      else
+        {
+          internal_fe_subface_values->reinit(cell, face_no, sub_face_no);
+          fe_face_values = internal_fe_subface_values.get();
+        }
+      if (sub_face_no_neighbor == numbers::invalid_unsigned_int)
+        {
+          internal_fe_face_values_neighbor->reinit(cell_neighbor,
+                                                   face_no_neighbor);
+          fe_face_values_neighbor = internal_fe_face_values_neighbor.get();
+        }
+      else
+        {
+          internal_fe_subface_values_neighbor->reinit(cell_neighbor,
+                                                      face_no_neighbor,
+                                                      sub_face_no_neighbor);
+          fe_face_values_neighbor = internal_fe_subface_values_neighbor.get();
+        }
+
+      AssertDimension(fe_face_values->n_quadrature_points,
+                      fe_face_values_neighbor->n_quadrature_points);
+
+      const_cast<unsigned int &>(this->n_quadrature_points) =
+        fe_face_values->n_quadrature_points;
     }
-  if (sub_face_no_neighbor == numbers::invalid_unsigned_int)
+  else if (internal_hp_fe_face_values)
     {
-      internal_fe_face_values_neighbor.reinit(cell_neighbor, face_no_neighbor);
-      fe_face_values_neighbor = &internal_fe_face_values_neighbor;
+      // Same as if above, but when hp is enabled.
+      if (sub_face_no == numbers::invalid_unsigned_int)
+        {
+          internal_hp_fe_face_values->reinit(cell, face_no);
+          fe_face_values = &const_cast<FEFaceValues<dim, spacedim> &>(
+            internal_hp_fe_face_values->get_present_fe_values());
```

Comment:
Have you considered the consequence of having different FE (or FE orders) across an interface? More specifically, if the mapping or quadrature rule is different on either side of the interface then this may lead to some issues or inconsistencies. It looks like your test `fe_interface_values_12.cc` doesn't cover this case, and I reckon if the `q_collection` was expanded then the change in number of q-points would trigger an assert in the jump/average value evaluation. [Here](https://github.com/dealii/dealii/blob/master/source/meshworker/scratch_data.cc#L315-L327) is how we dealt with in ScratchData (the application case was DG FEM, I believe).

Reply:
Oh, I was indeed assuming that this was not supported. Thanks a lot for pointing me to `ScratchData`, I didn't check that.

Reply:
Fixed here https://github.com/dealii/dealii/blob/7fb474eaa881abafb89a25bf3582ae625d027c37/include/deal.II/fe/fe_interface_values.h#L2440-L2442